### PR TITLE
Fix Orchard is not asking for an email when the email claim is not available

### LINKED ISSUE #7697: Orchard is not asking for an email when the email claim is not available
When you use an external provider and configure registration to no ask email address, Orchard is not asking for an email and it create a user without email.

With this configuration

image

It creates a user like this

image

### PR DESCRIPTION
**Pull request by jptissot**
Fixes #7697
Also moved the User and Role scripts to a `IExternalLoginEventHandler`

### CODE REVIEW

File: src/OrchardCore.Modules/OrchardCore.Users/Handlers/ScriptExternalLoginEventHandler.cs [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+                var jsonSerializerSettings = new JsonSerializerSettings()
+                {
+                    ContractResolver = new CamelCasePropertyNamesContractResolver()
+                };
+                var script = $"js: function generateUsername(context) {{\n{registrationSettings.GenerateUsernameScript}\n}}\nvar context = {JsonConvert.SerializeObject(context, jsonSerializerSettings)};\ngenerateUsername(context);\nreturn context;";
+
+                dynamic evaluationResult = _scriptingManager.Evaluate(script, null, null, null);
+                if (evaluationResult?.userName != null)
+                {
+                        return evaluationResult.userName;
```
Comment (jptissot) [inline src/OrchardCore.Modules/OrchardCore.Users/Handlers/ScriptExternalLoginEventHandler.cs:49]: Formatting

File: src/OrchardCore.Modules/OrchardCore.Users/Handlers/ScriptExternalLoginEventHandler.cs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+            }
+            return String.Empty;
+        }
+
+        public async Task UpdateRoles(UpdateRolesContext context)
+        {
+            var loginSettings = (await _siteService.GetSiteSettingsAsync()).As<LoginSettings>();
+            if (loginSettings.UseScriptToSyncRoles)
+            {
+                var jsonSerializerSettings = new JsonSerializerSettings()
```
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.Users/Handlers/ScriptExternalLoginEventHandler.cs:60]: static

File: src/OrchardCore.Modules/OrchardCore.Users/Handlers/ScriptExternalLoginEventHandler.cs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        }
+
+        public async Task<string> GenerateUserName(string provider, IEnumerable<SerializableClaim> claims)
+        {
+            var registrationSettings = (await _siteService.GetSiteSettingsAsync()).As<RegistrationSettings>();
+
+            if (registrationSettings.UseScriptToGenerateUsername)
+            {
+                var context = new { userName = String.Empty, loginProvider = provider, externalClaims = claims };
+                var jsonSerializerSettings = new JsonSerializerSettings()
```
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.Users/Handlers/ScriptExternalLoginEventHandler.cs:40]: static

File: src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs [hunk 4]
```diff
@@ -760,60 +732,34 @@ public async Task<IActionResult> RemoveLogin(RemoveLoginViewModel model)
         private async Task<string> GenerateUsername(ExternalLoginInfo info)
         {
             var now = new TimeSpan(_clock.UtcNow.Ticks) - new TimeSpan(DateTime.UnixEpoch.Ticks);
-            var ret = string.Concat("u" + Convert.ToInt32(now.TotalSeconds).ToString());
+            var ret = String.Concat("u" + Convert.ToInt32(now.TotalSeconds).ToString());
```
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs:735]: so is this the default user name it generates?

How can total seconds be unique? Wouldn't it be safer for this to be a guid, but onlye generated if there was no result?

**Review by deanmarcussen — COMMENTED** [review]
I like it, need more time to understand, but think the move is good. makes it much easier to alter behaviour if the default script behaviour isn't suitable, or causing issues with AD mapping's being of.

File: src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs (hunk 4 above)
Comment (jptissot) [reply src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs:735]: I was thinking about that too, felt like there could easily be name collisions here on a high traffic website.

File: src/OrchardCore.Modules/OrchardCore.Users/Views/RegistrationSettings.Edit.cshtml [hunk 5]
```diff
@@ -129,12 +130,16 @@
...
                 'switch (context.loginProvider) {\n' +
                 '    case "AzureAD":\n' +
-                '        context.userName = "azad" + Date.now().toString();\n' +
+                '        context.userName = "azad" + uuid();\n' +
                 '        break;\n' +
                 '    default:\n' +
-                '        log("Warning", "Provider " + loginProvider + " was not handled");\n' +
+                '        log("Warning", "Provider " + context.loginProvider + " was not handled");\n' +
+                '        // Uncomment to generate a username as a uuid\n' +
+                '        // context.userName "u" + uuid();\n' +
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Users/Views/RegistrationSettings.Edit.cshtml:142]:
```suggestion
                '        // context.userName = "u" + uuid();\n' +
```

File: src/OrchardCore/OrchardCore.Users.Abstractions/Handlers/UpdateRolesContext.cs [hunk 6]
```diff
@@ -40,11 +40,11 @@ public UpdateRolesContext(IUser user, string loginProvider, IEnumerable<Serializ
         /// <summary>
         /// Gets the roles to be added to the user roles.
         /// </summary>
-        public string[] RolesToAdd { get; set; } = new string[0];
+        public List<string> RolesToAdd { get; set; } = new List<string>();
```
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.Users.Abstractions/Handlers/UpdateRolesContext.cs:43]:
```suggestion
        public List<string> RolesToAdd { get; } = new List<string>();
```

File: src/OrchardCore/OrchardCore.Users.Abstractions/Handlers/UpdateRolesContext.cs [hunk 7]
```diff
@@ -40,11 +40,11 @@ public UpdateRolesContext(IUser user, string loginProvider, IEnumerable<Serializ
...
 
         /// <summary>
         /// Gets the roles to be removed from the user roles.
         /// </summary>
-        public string[] RolesToRemove { get; set; } = new string[0];
+        public List<string> RolesToRemove { get; set; } = new List<string>();
```
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.Users.Abstractions/Handlers/UpdateRolesContext.cs:48]:
```suggestion
        public List<string> RolesToRemove { get; } = new List<string>();
```

**Review by sebastienros — APPROVED** [review]

**Review by sebastienros — APPROVED** [review]

File: src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs (hunk 4 above)
Comment (jptissot) [reply src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs]: @deanmarcussen do you agree on the change to a uuid?

File: src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs [hunk 8]
```diff
@@ -759,61 +731,34 @@ public async Task<IActionResult> RemoveLogin(RemoveLoginViewModel model)
 
         private async Task<string> GenerateUsername(ExternalLoginInfo info)
         {
-            var now = new TimeSpan(_clock.UtcNow.Ticks) - new TimeSpan(DateTime.UnixEpoch.Ticks);
-            var ret = string.Concat("u" + Convert.ToInt32(now.TotalSeconds).ToString());
+            var ret = String.Concat("u" + IdGenerator.GenerateId());
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs:734]:
```suggestion
            var ret = String.Concat("u", IdGenerator.GenerateId());
```

**Pull request merged**